mr/src/itest/java/org/elasticsearch/hadoop/VersionTest.java

```diff
@@ -42,12 +56,121 @@ public void testNormalize() throws Exception {
+
+        // File in original directory
+        File testFile = new File(privateTmpDir, "test");
+        createSafe(testFile);
+
+        // Same file but in linked dir
+        File linkedTestFile = new File(tmpDir, "test");
+
+        // Same file but referenced in an absolute, but non-canonical manner
+        File awkwardTestFile = new File(linkTest.getAbsolutePath() + "/private/tmp/./test");
+        File awkwardParentTestFile = new File(linkTest.getAbsolutePath() + "/private/tmp/other/../test");
+
+        // Assert paths are different but canonical resolves as the same
+        assertTrue(testFile.exists());
+        assertTrue(linkedTestFile.exists());
+        assertTrue(awkwardTestFile.exists());
+        assertTrue(awkwardParentTestFile.exists());
+
+        assertNotEquals(testFile.getAbsolutePath(), linkedTestFile.getAbsolutePath());
+        assertNotEquals(testFile.getAbsolutePath(), awkwardTestFile.getAbsolutePath());
+        assertNotEquals(testFile.getAbsolutePath(), awkwardParentTestFile.getAbsolutePath());
+
+        assertEquals(IOUtils.toCanonicalFilePath(testFile.toURI().toURL()), IOUtils.toCanonicalFilePath(linkedTestFile.toURI().toURL()));
+        assertEquals(IOUtils.toCanonicalFilePath(testFile.toURI().toURL()), IOUtils.toCanonicalFilePath(awkwardTestFile.toURI().toURL()));
+        assertEquals(IOUtils.toCanonicalFilePath(testFile.toURI().toURL()), IOUtils.toCanonicalFilePath(awkwardParentTestFile.toURI().toURL()));
+
+        URL testURL = new URL("jar:file:" + testFile.getAbsolutePath() + "!/some/package/Hello.class");
+        URL linkedTestURL = new URL("jar:file:" + linkedTestFile.getAbsolutePath() + "!/some/package/Hello.class");
+        URL awkwardTestURL = new URL("jar:file:" + awkwardTestFile.getAbsolutePath() + "!/some/package/Hello.class");
+        URL awkwardParentTestURL = new URL("jar:file:" + awkwardParentTestFile.getAbsolutePath() + "!/some/package/Hello.class");
+
+        assertNotEquals(testURL.toString(), linkedTestURL.toString());
+        assertNotEquals(testURL.toString(), awkwardTestURL.toString());
+        assertNotEquals(testURL.toString(), awkwardParentTestFile.toString());
+
+        assertEquals(IOUtils.toCanonicalFilePath(testURL), IOUtils.toCanonicalFilePath(linkedTestURL));
+        assertEquals(IOUtils.toCanonicalFilePath(testURL), IOUtils.toCanonicalFilePath(awkwardTestURL));
+        assertEquals(IOUtils.toCanonicalFilePath(testURL), IOUtils.toCanonicalFilePath(awkwardParentTestURL));
+    }
+
+    private void createSafe(File testFile) throws IOException {
+        if (!testFile.createNewFile()) {
+            throw new IOException("Could not create test file " + testFile.toString());
+        }
+    }
+
+    private void mkdirSafe(File dir) throws IOException {
+        if (!dir.mkdirs()) {
+            throw new IOException("Could not create " + dir + " folder");
+        }
+    }
+
+    private void linkDir(File privateTmpDir, File tmpDir) throws IOException {
```

Comment:
nit: try to avoid switching the order of arguments (e.g. tmpDir is second arg on this method, but first arg on the one it is wrapping)